Build Themes from source

0xazure opened the pull request:
Closes #47.
Closes #271.

This implements building of themes into a `themes.json` file under `Extensions/dist/page/` similar to how Extensions are currently built.

Themes have been split into their own separate `.css` files and live under `Themes/`.

`Extensions/dist/extensions.js` has been refactored and code common to building both Extensions and Themes has moved to `Extensions/dist/resources.js`.

This also implements a build task for building `themes.json`, adds Theme files to CSS linting, and recompiles Themes on change for `gulp server`.

### Gulp Tasks:

#### `gulp clean:themes`

Cleans previously built `Extensions/dist/page/themes.json` files.

#### `gulp build:themes`

Builds a `themes.json` file from source and outputs it to `Extensions/dist/page/themes.json`.

See also: `gulp clean:themes`

0xazure commented:
A number of themes make use of external resources for things like backgrounds.  How should we handle these assets?  Leave them as-is, or host them elsewhere (GitHub, Tumblr, etc)?

Orange Wall - http://www.studioxenix.com/seven/themes/themes/orange_wall/index.gif
Tex - /images/status_bg.png?2
Faux Kawaii - http://www.studioxenix.com/seven/themes/themes/faux_kawaii/index.gif
Simplicity - http://puaga.com/seven/themes/themes/simplicity/black_stripes_design_10.jpg
Flowers & Lace - http://25.media.tumblr.com/d4ce447ffaf5ed3429d2c56b67bb115c/tumblr_mnsod5Cpot1stvm9wo1_500.jpg
Mega Rainbows - http://i7.minus.com/ijJTTw79OdTRu.png
Nebula - http://farm8.staticflickr.com/7390/8718908101_4cd0e29183_o.jpg
Crocodile Green - http://media.tumblr.com/e36d38530c1106a037ccb7a348c9ea5d/tumblr_inline_mox4uctYE91qz4rgp.jpg
White - http://static.tumblr.com/gwc9rra/Lx3mqcrmf/transparent-pagination.png
Starry Nights - http://www.wallpapershdi.com/wallpaper-original/wallpapers/nebula-1668.jpg

/cc @hobinjk

ThePsionic commented:
I say host them on GitHub. A per-theme folder with theme assets?

0xazure commented:
I would tend to agree, gives us better control over the assets and prevents them from up and disappearing on us. My only concern is around copyright; we have no idea where most of these images are from so we'd have to investigate that.

nightpool commented:
putting them all in the repo seems super weird, maybe we could host them in
private posts on tumblr?

On Sun, Aug 23, 2015 at 12:25 PM 0xazure <notifications@github.com> wrote:

> I would tend to agree, gives us better control over the assets and
> prevents them from up and disappearing on us. My only concern is around
> copyright; we have no idea where most of these images are from so we'd have
> to investigate that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/new-xkit/XKit/pull/340#issuecomment-133880608>.
>

nightpool commented:
tumblr sometimes breaks images though I think.

On Sun, Aug 23, 2015 at 1:23 PM Evan G <eg1290@gmail.com> wrote:

> putting them all in the repo seems super weird, maybe we could host them
> in private posts on tumblr?
>
> On Sun, Aug 23, 2015 at 12:25 PM 0xazure <notifications@github.com> wrote:
>
>> I would tend to agree, gives us better control over the assets and
>> prevents them from up and disappearing on us. My only concern is around
>> copyright; we have no idea where most of these images are from so we'd have
>> to investigate that.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/new-xkit/XKit/pull/340#issuecomment-133880608>.
>>
>

0xazure commented:
Why is it weird to put assets we rely on in the repo?  We'd be committing all of the extension icons too were it not for the fact that we base64 encode and inline them.

hobinjk commented:
@0xazure I agree with hosting them somewhere in the repo as assets. We should add a disclaimer README in the directory with the images stating how we don't know the copyright status of any of them.

0xazure commented:
Okay, I'll get to work on incorporating that change.

I'll also see what I can track down in terms of original sources for the files.  I'm not going to hold my breath though, who knows where any of them are from.

0xazure commented:
Updated themes to load images from our `gh-pages` GitHub URL.  I'll need a sanity check on the URLs and to make sure I haven't missed any.

Removed _Flowers & Lace_ and _Starry Nights_ because their remote assets are no longer available.  They were both developed by 3rd parties and not XKit itself.

I also had a quick look into the origins of the resources I've added; they all appear to be free-to-use or have been proliferated so widely that it would be unreasonable to expect copyright enforcement.  I don't anticipate any issues with their use.

0xazure commented:
:bell: Pinging here.  This PR is a blocker for a couple other improvements I have lined up for our `gulp` tasks, I would really appreciate a review.

hobinjk commented:
`create-extensions-dist.sh` should be updated to also call build:themes

hobinjk commented on Extensions/dist/themes.js:
```diff
@@ -0,0 +1,99 @@
+	themeFiles.forEach(function(fileName) {
+		// Look only for css files
+		var fileNameParts = fileName.split('.');
+		if (fileNameParts.length !== 2) {
```
gulp build:themes is generating an empty themes.json file and it looks like this is the culprit. I believe that because each theme is in its own directory this is incorrect in assuming that there will be a css file.

0xazure commented:
Updated `create-extensions-dist.sh` to build themes from source before git committing them to `gh-pages`.

Updated how we find source files under `Extensions/` and `Themes/` to build from.  Excellent catch, @hobinjk, `themes.json` was empty because we couldn't find any themes to process after I moved everything into folders.  We now [`glob`](https://github.com/isaacs/node-glob) for source files relative to `Extensions/` and `Themes/` and pass those paths along for processing.

hobinjk commented:
LGTM, great work!
@homu r+

homu commented:
:pushpin: Commit 7404fb4 has been approved by `hobinjk`

<!-- @homu r=hobinjk 7404fb4d232d9a45e6d755e8f2e3b8700a359e7e -->

homu commented:
:hourglass: Testing commit 7404fb4 with merge f04d87e...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/new-xkit/XKit/builds/78516315)
